apps/emqx/src/emqx_cm.erl

```diff
@@ -201,11 +201,27 @@ unregister_channel(ClientId) when ?IS_CLIENTID(ClientId) ->
 do_unregister_channel({_ClientId, ChanPid} = Chan) ->
     ok = emqx_cm_registry:unregister_channel(Chan),
     true = ets:delete(?CHAN_CONN_TAB, Chan),
-    true = ets:delete(?CHAN_INFO_TAB, Chan),
+    ok = maybe_transfer_channel_info_to_persistent_session(Chan),
     ets:delete_object(?CHAN_TAB, Chan),
     ok = emqx_hooks:run('cm.channel.unregistered', [ChanPid]),
     true.
 
+%% Takes a "last snapshot" of the channel info and statistics and hands it over to the
+%% persistent session, if any.  If persistence is not enabled or the session doesn't
+%% exist, it does nothing and simply drops the entry.
+maybe_transfer_channel_info_to_persistent_session({ClientId, _ChanPid} = Chan) ->
+    PersistenceEnabled = emqx_persistent_message:is_persistence_enabled(),
+    case ets:take(?CHAN_INFO_TAB, Chan) of
+        [{Chan, ChanInfo, Stats}] when PersistenceEnabled ->
+            emqx_persistent_session_ds_state:set_offline_info_now(
```

Comment:
Only the channel process owning the session can modify the durable session state. The only exception to this rule is GC process that drops durable states.
We cannot let arbitrary processes access this API for writing, as this would inevitably lead to session state corruption bugs.